virtual_rainforest/models/soil/carbon.py

```diff
@@ -0,0 +1,203 @@
+        soil_moisture: NDArray[np.float32],
+        soil_temp: NDArray[np.float32],
+        percent_clay: NDArray[np.float32],
+    ) -> NDArray[np.float32]:
+        """Calculates net rate of LMWC association with soil minerals.
+
+        Following Abramoff et al. (2018), mineral adsorption of carbon is controlled by
+        a Langmuir saturation function. At present, binding affinity and Q_max are
+        recalculated on every function called based on pH, bulk density and clay
+        content. Once a decision has been reached as to how fast pH and bulk density
+        will change (if at all), this calculation may need to be moved elsewhere.
+
+        Args:
+            pH: pH values for each soil grid cell
+            bulk_density: bulk density values for each soil grid cell
+            soil_moisture: soil moisture for each soil grid cell
+            soil_temp: soil temperature for each soil grid cell
+            percent_clay: Percentage clay for each soil grid cell
+
+        Returns:
+            lmwc_to_maom: The net flux from LMWC to MAOM
+        """
+
+        # This expression is drawn from (Mayes et al. (2012))
+        binding_affinity = 10.0 ** (
+            BINDING_WITH_PH["slope"] * pH + BINDING_WITH_PH["intercept"]
+        )
+
+        # This expression is also drawn from Mayes et al. (2012)
+        # Original paper also depends on Fe concentration, but we are ignoring this for
+        # now
+        Q_max = bulk_density * 10 ** (
+            MAX_SORPTION_WITH_CLAY["slope"] * np.log10(percent_clay)
+            + MAX_SORPTION_WITH_CLAY["intercept"]
+        )
+
+        # Using the above calculate the equilibrium MAOM pool
+        equib_maom = (binding_affinity * Q_max * self.lmwc) / (
+            1 + self.lmwc * binding_affinity
+        )
+
+        # Find scalar factors that multiple rates
+        temp_scalar = convert_temperature_to_scalar(soil_temp)
+        moist_scalar = convert_moisture_to_scalar(soil_moisture)
+
+        return temp_scalar * moist_scalar * self.lmwc * (equib_maom - self.maom) / Q_max
+
+
+def convert_temperature_to_scalar(
+    soil_temp: NDArray[np.float32],
+) -> NDArray[np.float32]:
+    """Convert soil temperature into a factor to multiply rates by.
+
+    This form is used in Abramoff et al. (2018) to minimise differences with the
+    CENTURY model. We very likely want to define our own functional form here. I'm
+    also a bit unsure how this form was even obtained, so further work here is very
+    needed.
+
+    Args:
+       soil_temp: soil temperature for each soil grid cell
+    """
+
+    # This expression is drawn from Abramoff et al. (2018)
+    numerator = TEMP_SCALAR["t_2"] + (TEMP_SCALAR["t_3"] / np.pi) * np.arctan(
+        np.pi * (soil_temp - TEMP_SCALAR["t_1"])
+    )
+    denominator = TEMP_SCALAR["t_2"] + (TEMP_SCALAR["t_3"] / np.pi) * np.arctan(
+        np.pi * TEMP_SCALAR["t_4"] * (TEMP_SCALAR["ref_temp"] - TEMP_SCALAR["t_1"])
+    )
```

Comment:
There are repeated dictionary lookups here. I don't know if this is a big deal (and I probably do something similar in `pyrealm`!) but you could do:

```
TEMP_SCALAR = np.array([15.4, 11.75,  29.7,  0.031,  30.0])
```

And then:

```python
t_1, t_2, t_3, t_4, ref_temp = TEMP_SCALAR

# This expression is drawn from Abramoff et al. (2018)
numerator = t_2 + (t_3 / np.pi) * np.arctan(np.pi * (soil_temp - t_1))
denominator = t_2 + (t_3 / np.pi) * np.arctan(np.pi * t_4 * (ref_temp - t_1))
```